From 302d0934b1d62f4961bd69d3ce9031df56c61c70 Mon Sep 17 00:00:00 2001
From: Murala S Swamy <ssmurala@habana.ai>
Date: Tue, 05 Nov 2024 14:54:56 +0200
Subject: [PATCH] [FSW-8694] fs1: control slave spi fuse offset init

To avoid undefined behaviour incase of d2d
gateway failure, key index flag from boot
pointer is used as slave fake fuse control
switch. This key index is unused in fake
fuse boot pointer, and if it set to 255
master will not initialize slave fake fuse
offsets.

Change-Id: I13b8d0855be2e0f4a2387c70c2e25c910c801977
---

diff --git a/common/boot_desc.c b/common/boot_desc.c
index c900a30..d440120 100644
--- a/common/boot_desc.c
+++ b/common/boot_desc.c
@@ -395,17 +395,24 @@
 	}
 
 	if (die_id_index == MAX_DIE_COUNT) {
-		for(index = 0; index < fake_fuse_boot_ptr->count; index++) {
+		if (fake_fuse_boot_ptr->key_idx == DISABLE_SLAVE_FAKE_FUSE) {
 			fake_fuse_img_desc = (struct hl_fw_img_desc *)(spi_manifest_consts.dest_addr_arc +
-				offset_array_base[index]);
-
-			/* Fake fuse bin order controlled by enum fake_fuse_bin_index */
-			if (((fake_fuse_img_desc->magic.type - 1) != (uint8_t)index) ||
-				(fake_fuse_img_desc->magic.type > MAX_DIE_COUNT))
-				return BOOT_ROM_FAIL;
-
-			fake_fuse_offset_array[index] = fake_fuse_base +
+				offset_array_base[FAKE_FUSE_IO_D0]);
+			fake_fuse_offset_array[FAKE_FUSE_IO_D0] = fake_fuse_base +
 				fake_fuse_img_desc->spi_offset[manifest_idx];
+		} else {
+			for(index = 0; index < fake_fuse_boot_ptr->count; index++) {
+				fake_fuse_img_desc = (struct hl_fw_img_desc *)(spi_manifest_consts.dest_addr_arc +
+					offset_array_base[index]);
+
+				/* Fake fuse bin order controlled by enum fake_fuse_bin_index */
+				if (((fake_fuse_img_desc->magic.type - 1) != (uint8_t)index) ||
+					(fake_fuse_img_desc->magic.type > MAX_DIE_COUNT))
+					return BOOT_ROM_FAIL;
+
+				fake_fuse_offset_array[index] = fake_fuse_base +
+					fake_fuse_img_desc->spi_offset[manifest_idx];
+			}
 		}
 	} else {
 		fake_fuse_img_desc = (struct hl_fw_img_desc *)(spi_manifest_consts.dest_addr_arc +
diff --git a/common/boot_desc.h b/common/boot_desc.h
index 5d971b6..aa4a91f 100644
--- a/common/boot_desc.h
+++ b/common/boot_desc.h
@@ -15,6 +15,8 @@
 #define INV_VER2		((uint32_t) -1)
 #define INVALID_DESC_ID		2
 
+#define DISABLE_SLAVE_FAKE_FUSE		255
+
 #define MAX_DIE_COUNT		(FS1_NUM_IO_DIES + FS1_NUM_COMPUTE_DIES)
 
 struct boot_desc {
